File: src/main/java/ordilov/lissn/like/infrastructure/LikeStoreImpl.java
package ordilov.lissn.like.infrastructure;

import lombok.RequiredArgsConstructor;
import ordilov.lissn.like.domain.LikeStore;
import ordilov.lissn.like.domain.LikedPlaylist;
import ordilov.lissn.like.domain.LikedTrack;
import org.springframework.stereotype.Repository;

@Repository
@RequiredArgsConstructor
public class LikeStoreImpl implements LikeStore {

  private final LikedTrackRepository likedTrackRepository;
  private final LikedPlaylistRepository likedPlaylistRepository;

  @Override
  public LikedTrack store(LikedTrack likedTrack) {
    return likedTrackRepository.save(likedTrack);
  }

  @Override
  public LikedPlaylist store(LikedPlaylist likedPlaylist) {
    return likedPlaylistRepository.save(likedPlaylist);
  }

  @Override
  public void delete(LikedTrack likedTrack) {
    likedTrack.getTrack().removeLikeCount();
    likedTrackRepository.delete(likedTrack);
  }

  @Override
  public void delete(LikedPlaylist likedPlaylist) {
    likedPlaylist.getPlaylist().removeLikeCount();
    likedPlaylistRepository.delete(likedPlaylist);
  }
}

File: src/main/java/ordilov/lissn/like/infrastructure/LikeStoreImpl.kt
package ordilov.lissn.like.infrastructure

import ordilov.lissn.like.domain.LikeStore
import ordilov.lissn.like.domain.LikedPlaylist
import ordilov.lissn.like.domain.LikedTrack
import org.springframework.stereotype.Repository

@Repository
class LikeStoreImpl(
    val likedTrackRepository: LikedTrackRepository,
    val likedPlaylistRepository: LikedPlaylistRepository
) : LikeStore {

    override fun store(likedTrack: LikedTrack): LikedTrack {
        return likedTrackRepository.save(likedTrack)
    }

    override fun store(likedPlaylist: LikedPlaylist): LikedPlaylist {
        return likedPlaylistRepository.save(likedPlaylist)
    }

    override fun delete(likedTrack: LikedTrack) {
        likedTrack.track.removeLikeCount()
        likedTrackRepository.delete(likedTrack)
    }

    override fun delete(likedPlaylist: LikedPlaylist) {
        likedPlaylist.playlist?.decreaseLikeCount()
        likedPlaylistRepository.delete(likedPlaylist)
    }
}